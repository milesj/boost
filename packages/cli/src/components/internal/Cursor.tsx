import React from 'react';
import { Text } from 'ink';
import { Style } from '../Style';

export interface CursorProps {
	focused?: boolean;
	hideCursor?: boolean;
	position: number;
	value: string;
}

export function Cursor({ focused, hideCursor, position, value }: CursorProps) {
	if (!focused) {
		return <Text>{value}</Text>;
	}

	return (
		<Text>
			{value.slice(0, position)}
			{hideCursor ? (
				value[position]
			) : (
				<Style inverted type="info">
					{value[position] || ' '}
				</Style>
			)}
			{value.slice(position + 1)}
		</Text>
	);
}
