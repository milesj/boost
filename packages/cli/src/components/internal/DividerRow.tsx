import React from 'react';
import { Box } from 'ink';
import { Style } from '../Style';

export interface DividerRowProps {
	label: NonNullable<React.ReactNode>;
}

export function DividerRow({ label }: DividerRowProps) {
	return (
		<Box marginLeft={2}>
			<Style type="muted">{label || '────'}</Style>
		</Box>
	);
}
