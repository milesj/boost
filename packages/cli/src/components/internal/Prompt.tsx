import React, { useState } from 'react';
import { Box, useInput, Key } from 'ink';
import { figures } from '@boost/terminal';
import { Label } from './Label';
import Style from '../Style';

export type KeyInput = Key;

export interface PromptProps<T> {
  defaultValue?: T;
  label: NonNullable<React.ReactNode>;
  focused?: boolean;
  prefix?: string;
  validate?: (value: T) => void;
}

export interface InternalPromptProps<T> extends PromptProps<T> {
  afterLabel?: React.ReactElement;
  beforeLabel?: React.ReactElement;
  children?: React.ReactNode;
  onBackspace?: (key: KeyInput) => void;
  onDelete?: (key: KeyInput) => void;
  onEscape?: (key: KeyInput) => void;
  onInput?: (value: string, key: KeyInput) => void;
  onKeyDown?: (key: KeyInput) => void;
  onKeyLeft?: (key: KeyInput) => void;
  onKeyRight?: (key: KeyInput) => void;
  onKeyUp?: (key: KeyInput) => void;
  onPageDown?: (key: KeyInput) => void;
  onPageUp?: (key: KeyInput) => void;
  onReturn?: (key: KeyInput) => void;
  onTab?: (key: KeyInput) => void;
  value: T;
}

export function Prompt<T>({
  afterLabel,
  beforeLabel,
  children,
  focused,
  label,
  onBackspace,
  onDelete,
  onEscape,
  onInput,
  onKeyDown,
  onKeyLeft,
  onKeyRight,
  onKeyUp,
  onPageDown,
  onPageUp,
  onReturn,
  onTab,
  prefix = '?',
  validate,
  value,
}: InternalPromptProps<T>) {
  const [error, setError] = useState<Error | null>(null);
  const [submitted, setSubmitted] = useState(false);

  useInput(
    // eslint-disable-next-line complexity
    (input, key) => {
      if (key.upArrow) {
        onKeyUp?.(key);
      } else if (key.downArrow) {
        onKeyDown?.(key);
      } else if (key.leftArrow) {
        onKeyLeft?.(key);
      } else if (key.rightArrow) {
        onKeyRight?.(key);
      } else if (key.pageUp) {
        onPageUp?.(key);
      } else if (key.pageDown) {
        onPageDown?.(key);
      } else if (key.return) {
        try {
          validate?.(value);
          onReturn?.(key);
          setError(null);
          setSubmitted(true);
        } catch (error: unknown) {
          if (error instanceof Error) {
            setError(error);
            setSubmitted(false);
          }
        }
      } else if (key.tab) {
        onTab?.(key);
      } else if (key.backspace) {
        onBackspace?.(key);
      } else if (key.delete) {
        onDelete?.(key);
      } else if (key.escape) {
        onEscape?.(key);
      } else {
        onInput?.(input, key);
        setSubmitted(false);
      }
    },
    { isActive: focused },
  );

  return (
    <Box flexDirection="column">
      <Box flexDirection="row">
        <Box>
          {submitted && !error && <Style type="success">{figures.tick}</Style>}
          {!submitted && error && <Style type="failure">{figures.cross}</Style>}
          {!submitted && !error && <Style type="info">{prefix}</Style>}
        </Box>

        {beforeLabel}

        <Box marginLeft={1} marginRight={1}>
          <Label>{label}</Label>
        </Box>

        {afterLabel}
      </Box>

      {error && (
        <Box marginLeft={2}>
          <Style type="failure">{error.message}</Style>
        </Box>
      )}

      {children && (
        <Box flexDirection="column" marginLeft={2}>
          {children}
        </Box>
      )}
    </Box>
  );
}
