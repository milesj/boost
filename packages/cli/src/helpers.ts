import { stripAnsi, style } from '@boost/terminal';
import { PrimitiveType } from '@boost/args/src';

export function applyMarkdown(text: string): string {
  return text.replace(/(?:\*|~|_){1,2}([^*~_]+)(?:\*|~|_){1,2}/giu, (match, body) => {
    if (match.startsWith('~')) {
      return style.strikethrough(body);
    } else if (match.startsWith('**') || match.startsWith('__')) {
      return style.bold(body);
    } else if (match.startsWith('*') || match.startsWith('_')) {
      return style.italic(body);
    }

    return match;
  });
}

export function formatValue(value: PrimitiveType): string {
  switch (typeof value) {
    case 'string':
      return `"${value}"`;
    default:
      return String(value);
  }
}

export function formatDescription(
  config: {
    choices?: PrimitiveType[];
    default?: PrimitiveType;
    description: string;
    deprecated?: boolean;
  },
  tags: string[] = [],
): string {
  let output = applyMarkdown(config.description);

  // Append choices after the description
  if (config.choices) {
    const choices = config.choices.map(
      choice => formatValue(choice) + (config.default === choice ? ' (default)' : ''),
    );

    output += ` (choices: ${choices.join(', ')})`;

    // Append default after description if no choices
  } else if (config.default) {
    output += ` (default: ${formatValue(config.default)})`;
  }

  // Tags go at the end of the description
  if (config.deprecated) {
    tags.unshift('deprecated');
  }

  if (tags.length > 0) {
    output += ' ';
    output += style.gray(`[${tags.join(', ')}]`);
  }

  return output;
}

export function formatType(
  config: {
    label?: string;
    multiple?: boolean;
    required?: boolean;
    type: string;
  },
  inline: boolean = false,
): string {
  let type = config.type + (config.multiple ? '[]' : '');

  if (inline) {
    if (config.label) {
      type = config.label;
    }

    type = config.required ? `<${type}>` : `[${type}]`;
  }

  return style.gray(type);
}

export function getLongestWidth(values: string[]): number {
  return values.reduce((sum, value) => {
    const text = stripAnsi(value);

    return text.length > sum ? text.length : sum;
  }, 0);
}
