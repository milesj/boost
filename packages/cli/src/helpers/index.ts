export * from './applyMarkdown';
export * from './applyStyle';
export * from './formatCommandCall';
export * from './formatDescription';
export * from './formatType';
export * from './formatValue';
export * from './getLongestWidth';
export * from './groupByCategory';
export * from './isArgvSize';
export * from './loadTheme';
export * from './mapCommandMetadata';
