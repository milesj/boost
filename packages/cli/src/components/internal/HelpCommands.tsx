import { useMemo, useRef } from 'react';
import { Box, Text } from 'ink';
import { stripAnsi } from '@boost/terminal';
import { SPACING_COL, SPACING_COL_WIDE, SPACING_ROW } from '../../constants';
import { formatCommandCall, formatDescription, groupByCategory } from '../../helpers';
import { msg } from '../../translate';
import type { Categories, CommandConfigMap } from '../../types';
import { Header } from '../Header';
import { Style } from '../Style';

export interface HelpCommandsProps {
	categories?: Categories;
	commands: CommandConfigMap;
}

export function HelpCommands({ categories, commands }: HelpCommandsProps) {
	const pathWidth = useRef(0);

	const items = useMemo(
		() =>
			Object.entries(commands).map(([path, command]) => {
				const pathCall = formatCommandCall(path, command);
				const pathCallStripped = stripAnsi(pathCall);

				if (pathCallStripped.length > pathWidth.current) {
					pathWidth.current = pathCallStripped.length;
				}

				return {
					...command,
					name: path,
					path,
					pathCall,
				};
			}),
		[commands],
	);

	const categorizedCommands = useMemo(
		() => groupByCategory(items, categories ?? {}),
		[categories, items],
	);

	const categoryCount = Object.keys(categorizedCommands).length;

	return (
		<>
			<Header label={msg('cli:labelCommands')} />

			{Object.entries(categorizedCommands).map(([key, category], index) => (
				<Box key={key} flexDirection="column">
					{!!category.name && categoryCount > 1 && (
						<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
							<Style bold type="none">
								{category.name}
							</Style>
						</Box>
					)}

					{category.items.map((config) => {
						const desc = formatDescription(config);

						return (
							<Box key={key + config.path} flexDirection="row" paddingLeft={SPACING_COL}>
								<Box flexGrow={0} width={pathWidth.current + SPACING_COL_WIDE}>
									<Text>{config.pathCall}</Text>
								</Box>

								<Box flexGrow={1}>
									<Text wrap="wrap">{desc}</Text>
								</Box>
							</Box>
						);
					})}
				</Box>
			))}
		</>
	);
}
