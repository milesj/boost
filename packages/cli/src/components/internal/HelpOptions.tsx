import { useMemo, useRef } from 'react';
import { Box, Text } from 'ink';
import { OptionConfig, OptionConfigMap } from '@boost/args';
import { stripAnsi } from '@boost/terminal';
import { SPACING_COL, SPACING_COL_WIDE, SPACING_ROW } from '../../constants';
import { formatDescription, formatType, groupByCategory } from '../../helpers';
import { msg } from '../../translate';
import { Categories } from '../../types';
import { Header } from '../Header';
import { Style } from '../Style';

function gatherOptionTags(config: OptionConfig): string[] {
	const tags: string[] = [];

	if (config.count) {
		tags.push(msg('cli:tagCount'));
	}

	if (config.multiple) {
		tags.push(msg('cli:tagMultiple') + (config.arity ? `(${config.arity})` : ''));
	}

	return tags;
}

export interface HelpOptionsProps {
	categories?: Categories;
	options: OptionConfigMap;
}

export function HelpOptions({ categories, options }: HelpOptionsProps) {
	const longWidth = useRef(0);
	const shortWidth = useRef(0);

	const items = useMemo(
		() =>
			Object.entries(options).map(([name, option]) => {
				const shortLabel = option.short ? `-${option.short},` : '';
				let longLabel = `--${option.default === true ? 'no-' : ''}${name}`;

				if (option.type !== 'boolean') {
					longLabel += ' ';
					longLabel += formatType(option);
				}

				const longLabelStripped = stripAnsi(longLabel);
				const shortLabelStripped = stripAnsi(shortLabel);

				if (longLabelStripped.length > longWidth.current) {
					longWidth.current = longLabelStripped.length;
				}

				if (shortLabelStripped.length > shortWidth.current) {
					shortWidth.current = shortLabelStripped.length;
				}

				return {
					...option,
					longLabel,
					name,
					shortLabel,
				};
			}),
		[options],
	);

	const categorizedOptions = useMemo(
		() => groupByCategory(items, categories ?? {}),
		[categories, items],
	);
	const categoryCount = Object.keys(categorizedOptions).length;
	const showShortColumn = shortWidth.current > 0;

	return (
		<>
			<Header label={msg('cli:labelOptions')} />

			{Object.entries(categorizedOptions).map(([key, category], index) => (
				<Box key={key} flexDirection="column">
					{!!category.name && categoryCount > 1 && (
						<Box marginTop={index === 0 ? 0 : SPACING_ROW} paddingLeft={SPACING_COL}>
							<Style bold type="none">
								{category.name}
							</Style>
						</Box>
					)}

					{category.items.map((config) => (
						<Box key={key + config.name} flexDirection="row" paddingLeft={SPACING_COL}>
							{showShortColumn && (
								<Box flexGrow={0} width={shortWidth.current + SPACING_COL}>
									<Text>{config.shortLabel}</Text>
								</Box>
							)}

							<Box flexGrow={0} width={longWidth.current + SPACING_COL_WIDE}>
								<Text>{config.longLabel}</Text>
							</Box>

							<Box flexGrow={1}>
								<Text wrap="wrap">{formatDescription(config, gatherOptionTags(config))}</Text>
							</Box>
						</Box>
					))}
				</Box>
			))}
		</>
	);
}
