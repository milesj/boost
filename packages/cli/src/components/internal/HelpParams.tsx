import React, { useMemo } from 'react';
import { Box, Text } from 'ink';
import { ParamConfig, ParamConfigList } from '@boost/args';
import { SPACING_COL, SPACING_COL_WIDE } from '../../constants';
import { formatDescription, formatType, getLongestWidth } from '../../helpers';
import { msg } from '../../translate';
import { CommandConfig } from '../../types';
import { Header } from '../Header';

function gatherParamTags(config: ParamConfig): string[] {
	const tags: string[] = [];

	if (config.required) {
		tags.push(msg('cli:tagRequired'));
	}

	return tags;
}

export interface HelpParamsProps {
	config?: CommandConfig;
	params: ParamConfigList;
}

export function HelpParams({ config, params }: HelpParamsProps) {
	const labels = useMemo(
		() => params.map((p, index) => `${p.label ?? index} ${formatType(p)}`),
		[params],
	);
	const labelWidth = getLongestWidth(labels);
	const allowVariadic = config?.allowVariadicParams;

	return (
		<>
			<Header label={msg('cli:labelParams')} />

			{params.map((param, index) => {
				if (param.hidden) {
					return null;
				}

				const desc = param ? formatDescription(param, gatherParamTags(param)) : '';

				return (
					<Box key={`${labels[index]}-${index}`} flexDirection="row" paddingLeft={SPACING_COL}>
						<Box flexGrow={0} width={labelWidth + SPACING_COL_WIDE}>
							<Text>{labels[index]}</Text>
						</Box>

						<Box flexGrow={1}>
							<Text wrap="wrap">{desc}</Text>
						</Box>
					</Box>
				);
			})}

			{allowVariadic && (
				<Box paddingLeft={SPACING_COL}>
					<Text>
						{`â€¦${typeof allowVariadic === 'string' ? allowVariadic : ''}`}{' '}
						{formatType({
							multiple: true,
							type: 'string',
						})}
					</Text>
				</Box>
			)}
		</>
	);
}
