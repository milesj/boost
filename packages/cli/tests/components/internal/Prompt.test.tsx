import React from 'react';
import { Box, Text } from 'ink';
import { render } from 'ink-testing-library';
import { Prompt } from '../../../src/components/internal/Prompt';
import { KEYS } from '../../helpers';

// http://ascii-table.com/ansi-escape-sequences-vt-100.php
describe('Prompt', () => {
	it('renders default state', () => {
		const { lastFrame } = render(<Prompt label="Label" value="" />);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders content before and after label', () => {
		const { lastFrame } = render(
			<Prompt
				beforeLabel={<Text>Before</Text>}
				label="Label"
				afterLabel={<Text>After</Text>}
				value=""
			/>,
		);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders content below label when focused', () => {
		const { lastFrame } = render(
			<Prompt focused label="Label" value="">
				<Box>
					<Text>Foo</Text>
				</Box>
				<Box>
					<Text>Bar</Text>
				</Box>
				<Box>
					<Text>Baz</Text>
				</Box>
			</Prompt>,
		);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('doesnt render content below label when not focused', () => {
		const { lastFrame } = render(
			<Prompt label="Label" value="">
				<Box>
					<Text>Foo</Text>
				</Box>
				<Box>
					<Text>Bar</Text>
				</Box>
				<Box>
					<Text>Baz</Text>
				</Box>
			</Prompt>,
		);

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders error below label and above content', async () => {
		const { lastFrame, stdin } = render(
			<Prompt
				focused
				label="Label"
				value=""
				validate={() => {
					throw new Error('Failed validation');
				}}
			>
				<Box>
					<Text>Foo</Text>
				</Box>
			</Prompt>,
		);

		await delay();
		stdin.write(KEYS.return);
		await delay();

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders success after return submission', async () => {
		const { lastFrame, stdin } = render(
			<Prompt label="Label" value="" onReturn={() => true}>
				<Box>
					<Text>Foo</Text>
				</Box>
			</Prompt>,
		);

		await delay();
		stdin.write(KEYS.return);
		await delay();

		expect(lastFrame()).toMatchSnapshot();
	});

	it('renders success after input submission', async () => {
		const { lastFrame, stdin } = render(
			<Prompt label="Label" value="" onInput={() => true}>
				<Box>
					<Text>Foo</Text>
				</Box>
			</Prompt>,
		);

		await delay();
		stdin.write('a');
		await delay();

		expect(lastFrame()).toMatchSnapshot();
	});

	it('calls `onKeyUp` for up arrow', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onKeyUp={spy} />);

		await delay();
		stdin.write(KEYS.up);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onKeyDown` for down arrow', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onKeyDown={spy} />);

		await delay();
		stdin.write(KEYS.down);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onKeyRight` for right arrow', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onKeyRight={spy} />);

		await delay();
		stdin.write(KEYS.right);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onKeyLeft` for left arrow', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onKeyLeft={spy} />);

		await delay();
		stdin.write(KEYS.left);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onPageUp` for page up', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onPageUp={spy} />);

		await delay();
		stdin.write(KEYS.pageUp);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onPageDown` for page down', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onPageDown={spy} />);

		await delay();
		stdin.write(KEYS.pageDown);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onTab` for tab', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onTab={spy} />);

		await delay();
		stdin.write(KEYS.tab);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onBackspace` for backspace', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onBackspace={spy} />);

		await delay();
		stdin.write(KEYS.backspace);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onDelete` for delete', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onDelete={spy} />);

		await delay();
		stdin.write(KEYS.delete);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onReturn` for return/enter', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onReturn={spy} />);

		await delay();
		stdin.write(KEYS.return);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onEscape` for escape', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onEscape={spy} />);

		await delay();
		stdin.write(KEYS.escape);
		await delay();

		expect(spy).toHaveBeenCalled();
	});

	it('calls `onInput` for standard characters', async () => {
		const spy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onInput={spy} />);

		await delay();
		stdin.write('a');
		await delay();

		expect(spy).toHaveBeenCalledWith('a', expect.any(Object));
	});

	it('calls `onSpace` (if provided) for space character', async () => {
		const spaceSpy = jest.fn();
		const inputSpy = jest.fn();
		const { stdin } = render(
			<Prompt label="Label" value="" onSpace={spaceSpy} onInput={inputSpy} />,
		);

		await delay();
		stdin.write(' ');
		await delay();

		expect(spaceSpy).toHaveBeenCalledWith(expect.any(Object));
		expect(inputSpy).not.toHaveBeenCalled();
	});

	it('calls `onInput` for space character if `onSpace` is not provided', async () => {
		const inputSpy = jest.fn();
		const { stdin } = render(<Prompt label="Label" value="" onInput={inputSpy} />);

		await delay();
		stdin.write(' ');
		await delay();

		expect(inputSpy).toHaveBeenCalledWith(' ', expect.any(Object));
	});
});
