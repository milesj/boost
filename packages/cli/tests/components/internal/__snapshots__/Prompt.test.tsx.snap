// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt doesnt render content below label when not focused 1`] = `"[36m?[39m [1mLabel[22m"`;

exports[`Prompt renders content before and after label 1`] = `"[36m?[39mBefore [1mLabel[22m After"`;

exports[`Prompt renders content below label when focused 1`] = `
"[36m?[39m [1mLabel[22m
  Foo
  Bar
  Baz"
`;

exports[`Prompt renders default state 1`] = `"[36m?[39m [1mLabel[22m"`;

exports[`Prompt renders error below label and above content 1`] = `
"[31mâœ–[39m [1mLabel[22m
  [31mFailed validation[39m
  Foo"
`;

exports[`Prompt renders success after input submission 1`] = `"[32mâœ”[39m [1mLabel[22m"`;

exports[`Prompt renders success after return submission 1`] = `"[32mâœ”[39m [1mLabel[22m"`;
